docs: fips+stig documentation release branch (#1911)
canonical/k8s-snap: docs/canonicalk8s/snap/howto/install/disa-stig.md

> # How to install {{ product }} with DISA STIG hardening

Reviewer: @nhennigan more a thought for the future, but maybe we should recast the "install" section as a "deployment" section. Really we are doing the install in FIPS and doing the bootstrap and DISA STIG in this part. I think the key point is that you can't rebootstrap. So any guide that requires bootstrapping should go in this section and you can only pick one of them. We could also call it "Bootstrap". We can discuss more later; out of scope for this PR I think.

Security Technical Implementation Guides (STIGs) are developed by the Defense
Information System Agency (DISA) and are comprehensive frameworks of security
requirements designed to protect U.S. Department of Defense (DoD) systems and
networks from cybersecurity threats.

The [Kubernetes STIG] contains guidelines on how to check and remediate various
potential security concerns for a Kubernetes deployment, both on the host and
within the cluster itself. This guide goes through the steps needed to deploy
{{product}} with our dedicated security configuration files to meet DISA STIG
hardening requirements.

## Prerequisites

This guide assumes the following:

- Ubuntu machine with at least 4GB of RAM and 30 GB disk storage
- You have root or sudo access to the machine
- Internet access on the machine
- You have Ubuntu Pro enabled on your system. For more information, see
  [Ubuntu Pro documentation]
- You have FIPS enabled on your machine. See our [FIPS installation guide] for

> guidance

Reviewer: ```suggestion

## Install {{product}} in FIPS mode

DISA STIG compliance also incorporates FIPS compliance. Start by following our [FIPS installation guide], but stop after installing {{product}} and do not follow the steps bootstrap the cluster. Instead continue here as there additional steps and configuration needed for DISA STIG compliance. The guide also covers enabling Ubuntu Pro, which is needed for some of the steps below.
```

I think this needs a little more explanation and might make sense in its own section vs being just a bullet in the pre-requesitte list. Ubuntu Pro really shouldn't be its own bullet since it's part of the FIPS guide.

## DISA STIG host compliance

DISA STIG host compliance is achieved by running the [USG tool] that is also
part of the PRO tool set. To install the USG tool:

```
sudo pro enable usg
sudo apt update
sudo apt install usg
```

To generate a compliance audit report (without applying changes):

```
sudo usg audit disa_stig
```

```{warning}
The following command applies rule [V-270714] which will cause issues using
accounts with an empty password to access your machine. To avoid being locked

> out use an RSA key to access your machine.

Reviewer: Removing second sentence works for me. Good point re types of keys supported; I suppose FIPS supports more than just RSA keys though? I don't think we need a note in the FIPS docs. This may vary for different environments/needs and isn't really in scope of this doc.

```

To automatically apply the recommended hardening changes:

```
sudo usg fix disa_stig
```

## Apply DISA STIG Kubernetes host rules

To comply with this guideline, the STIG templates we provide to bootstrap/join
nodes configure kubelet to run with the argument
`--protect-kernel-defaults=true`.

Configure the kernel as required for this setting by following the steps below:

```
sudo tee /etc/sysctl.d/99-kubelet.conf <<EOF
vm.overcommit_memory=1
vm.panic_on_oom=0
kernel.keys.root_maxbytes=25000000
kernel.keys.root_maxkeys=1000000
kernel.panic=10
kernel.panic_on_oops=1
EOF
sudo sysctl --system
```

```{note}
Please ensure that the configuration of `/etc/sysctl.d/99-kubelet.conf` is not
overridden by another higher order file.
```

> 

Reviewer: Also, it looks like the FW page was merged to main already a few weeks ago. Some issues I noticed though:

1. "Package forwarding" rather than "Packet forwarding"
2. Etcd port 2379 is missing from the instructions
3. The section about IP forwarding is confusing and maybe wrong. It says to edit sysctl.conf, but doesn't say to reboot and doesn't load the config file. But then it says, "Otherwise..." set the setting directly, but it never really said to reboot. See the doc for one way to handle this.

CC: @ktsakalozos

## Deploy and bootstrap STIG-compliant nodes

For your convenience, we have provided template configuration files that can be
used to configure Kubernetes service arguments that align with DISA STIG

> requirements.

Reviewer: ```suggestion
For your convenience, we have provided template configuration files that can be
used to configure the additional Kubernetes service arguments that are needed to fully align with DISA STIG
requirements.
```

### Bootstrap the first control-plane node

Install {{product}}:

```{literalinclude} ../../../_parts/install.md
:start-after: <!-- snap start -->
:end-before: <!-- snap end -->
```

Initialize the first control plane node with the necessary arguments for
DISA STIG compliance:

```
sudo k8s bootstrap --file /var/snap/k8s/common/etc/templates/disa-stig/bootstrap.yaml
sudo k8s status --wait-ready
```

Through this configuration file, the following rules are applied:

> | STIG | Summary |

Reviewer: @nhennigan thought for the future. I wonder if it would help to have a separate reference page that covers the available bootstrap templates and which settings are applied for each rather than having those extra details here.

Reviewer: Right, but still that's 3 different files and two lists of what they apply. It's okay to leave them on this page for now, but I think there's enough content already that a page on its own wouldn't hurt.

|------|---------|
|[V-242434] |Kubernetes Kubelet must enable kernel protection|
| [V-245541] |Kubernetes Kubelet must not disable timeouts|
| [V-242402], [V-242403], [V-242461], [V-242462], [V-242463], [V-242464], [V-242465] |The Kubernetes API Server must have an audit log configured|
| [V-254800] | Kubernetes must have a Pod Security Admission control file configured |
| [V-242400] | The Kubernetes API server must have Alpha APIs disabled|
| [V-242384] | The Kubernetes Scheduler must have secure binding |
| [V-242385] | The Kubernetes Controller Manager must have secure binding |

### Join additional control plane nodes

To join another control plane node to the cluster, first retrieve the join token
from an existing control plane node:

```
sudo k8s get-join-token <joining-node-hostname>
```

On the joining control plane node install {{product}}:

```{literalinclude} ../../../_parts/install.md
:start-after: <!-- snap start -->
:end-before: <!-- snap end -->
```

> and join the cluster with the DISA STIG control plane configuration file:

Reviewer: ```suggestion
On the additional control plane node, join the cluster with the DISA STIG control plane configuration file:
```
Again, the install will be done already as a part of the FIPS guide.

```
sudo k8s join-cluster --file=/var/snap/k8s/common/etc/templates/disa-stig/control-plane.yaml <join-token>
```

Through this configuration file, the same rules that were applied to the first
control plane node have also been applied.

### Join worker nodes

To join a worker node to the cluster, first retrieve the join token from an
existing control plane node:

```
sudo k8s get-join-token <joining-node-hostname> --worker
```

On the joining worker node, install {{product}}:

```{literalinclude} ../../../_parts/install.md
:start-after: <!-- snap start -->
:end-before: <!-- snap end -->
```

and join the cluster with the DISA STIG worker configuration file:

```
sudo k8s join-cluster --file=/var/snap/k8s/common/etc/templates/disa-stig/worker.yaml <join-token>
```

> Through this configuration file, the following rules are applied:

Reviewer: ```suggestion
This configuration file applies the following rules:
```

```

```{note}
According to rule [V-242393] and [V-242394] Kubernetes Worker Nodes must not have
sshd service running or enabled. The host STIG rule [V-270665] on the other hand
expects sshd to be installed on the host. To comply with both rules, leave SSH
installed, but disable the service.
```

## Alternative configurations

### Pod Security Admission control file

To comply with rule [V-254800], you must configure a Pod Security Admission
control file for your Kubernetes cluster. This file defines the Pod Security
Standards (PSS) that are enforced at the namespace level. The STIG templates we
provide to bootstrap/join nodes configure the Pod Security admission controller
to comply with these recommendations. For your own cluster, assess if:

- The default config at
`/var/snap/k8s/common/etc/configurations/pod-security-admission-baseline.yaml`
meets your needs or if it needs to be adjusted
- Using the more restrictive config we provide at
`/var/snap/k8s/common/etc/configurations/pod-security-admission-restricted.yaml`
is more suited to your needs
- You need to create your own audit policy based on the
[upstream instructions] and adjust the `--admission-control-config-file` path

> used when you bootstrap/join nodes.

Reviewer: This was adjusted from what's in the doc, but I found it less clear. Read literally, this says to assess whether any of the 3 options are needed, but doesn't say what to do. It also doesn't say anything about why someone would want to make these changes.

levels, and configuration options.

### Kubernetes API Server audit log

This applies to rules [V-242402], [V-242403], [V-242461], [V-242462],
[V-242463],[V-242464] and [V-242465]. The STIG templates we provide to
bootstrap/join nodes configures the Kubernetes
API servers audit settings and policy to comply with these recommendations.

The default level of logging may be impractical for some situations, in which
case the settings would need to be adjusted and an exception put in place. For
your own cluster, assess if:

- The default logging configuration set at
`/var/snap/k8s/common/etc/configurations/audit-policy.yaml`, which configures
logging of all (non-resource) events with request metadata, request body, and
response body is suitable for your needs or if it needs to be adjusted
- Using the audit policy we provide at
`/var/snap/k8s/common/etc/configurations/audit-policy-kube-system.yaml` which
configures the same level of logging but only for events in the kube-system
namespace is more suited to your needs
- You need to create your own audit policy based on the
[upstream audit instructions] and adjust the `--audit-policy-file` path used

> when you bootstrap/join nodes to use it.

Reviewer: Again, this literally says to make an assessment, but doesn't say to do anything. I would fix or revert to what we had in the doc.

In the default bootstrap configuration, Canonical Kubernetes does not enable
audit logging as it may incur
performance penalties in the form of increased disk I/O, which can lead to
slower response times and reduced overall cluster efficiency, especially under

> heavy workloads.

Reviewer: ```suggestion
```
This is the only place on this page where we explain why something isn't configured by default, so it's out of place. We either need to explain this everywhere or nowhere. I thought we were going to have this information in the DISA STIG audit reference page; that page needs a lot of work though.

## Manual checks

There are certain rules that must be completed manually in order to achieve
DISA STIG compliance.

- [V-242383]  User-managed resources must be created in dedicated namespaces so
you must manually inspect all services in the default namespace
- [V-242410], [V-242411], [V-242412], [V-242413] The Kubernetes API Server,
Scheduler, Controllers and etcd must enforce ports, protocols, and services
(PPS) that adhere to the Ports, Protocols, and Services Management Category
Assurance List (PPSM CAL). The {{product}} [ports and services] must be
manually reviewed in accordance with this policy
- [V-242414] The Kubernetes cluster must use non-privileged host ports for user
pods so you must manually inspect all pods in the default namespace
- [V-242415] Secrets in Kubernetes must not be stored as environment variables
so you must manually inspect all pods in the default namespace
- [V-242417] Kubernetes must separate user functionality

> so you must manually inspect all pods in the default namespace

Reviewer: ```suggestion
## Required Documentation

[V-242410], [V-242411], [V-242412], and [V-242413] requires that the Kubernetes API Server,
Scheduler, and Controllers as well as etcd enforce ports, protocols, and services
(PPS) that adhere to the Ports, Protocols, and Services Management Category
Assurance List (PPSM CAL). The {{product}} [ports and services] must be
manually audited in accordance with this policy. These ports, protocols, and services will need added to your specific PPSM list and the list will need updated anytime the list of ports, protocols, and services used by your cluster changes. For instance, this list will need to be updated each time a new service is exposed externally.

## Post-Deployment Requirements

In addition to the above deployment steps, there are some guidelines that must be followed by users and administrators throughout the life of the cluster in order to maintain DISA STIG compliance.

- [V-242383]  User-managed resources must be created in dedicated namespaces
- [V-242414] The Kubernetes cluster must use non-privileged host ports for user pods
- [V-242415] Secrets in Kubernetes must not be stored as environment variables
- [V-242417] Kubernetes must separate user functionality
```

I split this into two sections for clarity and because I thought more detail was needed for the PPSM CAL whereas less is needed for the other items. I don't think it's helpful to have the additional generic statement saying to audit those items without providing the details. 

Maybe it would be better to have the ongoing obligations documented on a separate page?

## Next steps

Please assess your cluster for compliance using the [DISA STIG auditing page].
Review all findings and apply any necessary remediations to be fully DISA STIG
compliant. Be aware that some rules need to be upheld when you add workloads

> to your cluster.

Reviewer: ```suggestion
## Auditing
If you would like to manually audit your cluster, see our [DISA STIG auditing page], where you will find detailed instructions and tools to evaluate your Kubernetes cluster against DISA STIG compliance requirements.
```
The way it was made it sound like a required step to be fully compliant. Really it's optional. A lot of me thinks we should just delete this section entirely. It feels out of place.
